feat: Additional audit log information for updating argocd applications (resolves #23130)

TomHellier opened the pull request:
When users update applications in the argocd ui, the audit log should be clear about what was changed in the application/applicationset. This change allows auditors to understand the difference in an update.

For example, if a user sets a helm parameter override, or points the application at a branch in the webui. Auditing should reflect that.

This information will not get included in the kubernetes event log, you can include very large differences in a `spec.source.helm.values` field, and it would clutter the kubernetes event log.

This change also adds logFields to the various calls to logAppEvent to allow future developers to add arbitrary contents to the audit logs.

Log Output (changed)
![image](https://github.com/user-attachments/assets/0140f0d0-982d-4f8b-b372-02b6b7938413)

Kubernetes Event Output (unchanged)
![image](https://github.com/user-attachments/assets/1ab1fafe-50d0-42b2-8a4e-7362751ada9e)
![image](https://github.com/user-attachments/assets/2b09e81e-6b9c-4052-a351-0e4c27b776c9)

Closes [ISSUE #23130]

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

TomHellier committed:
When users update applications in the argocd ui, the audit log should be clear about what was changed in the application spec. This change allows auditors to understand the difference in an update.

For example, if a user sets a helm parameter override, or points the application at a branch in the webui. Auditing should reflect that.

This information will also get included in the kubernetes event log, but I think that is ok as I wouldn't expect very large differences between applications.

Signed-off-by: Tom Hellier <me@tomhellier.com>

crenshaw-dev commented on server/application/application.go:
```diff
@@ -987,6 +991,23 @@ func (s *Server) updateApp(ctx context.Context, app *v1alpha1.Application, newAp
 	return nil, status.Errorf(codes.Internal, "Failed to update application. Too many conflicts")
 }
 
+func diffBetweenApplicationSpecs(a *v1alpha1.ApplicationSpec, b *v1alpha1.ApplicationSpec) (string, error) {
```
A couple suggestions:

1) I'd unmarshal the patch into a map[any]any and log it as its own field, so it's parseable by log tools
2) I'd exclude the patch from the event: some people put _huge_ stuff in their app spec

TomHellier replied:
Hi @crenshaw-dev  - thanks for the review.

I've pushed changes based on my understanding of your feedback, if you wouldn't mind another look.

TomHellier committed:
Add logFields into the audit and logging information, and make clear what are annotations vs logs

It looks like logFields were previously being dual used as annotations on the event, so splitting logFields into logFields and eventAnnotations, as we want to start using logFields to log additional information in the audit log. This could be expanded in future PRs to add more context to the audit logs.

Signed-off-by: Tom Hellier <me@tomhellier.com>

TomHellier commented on server/application/application.go:
```diff
@@ -2028,16 +2053,16 @@ func (s *Server) Sync(ctx context.Context, syncReq *application.ApplicationSyncR
 		partial = "partial "
 	}
-	var reason string
+	var action string
```
this is an "action", not a "reason".

argo.EventReasonOperationStarted is the reason, so makes sense to update this.

TomHellier commented on util/argo/audit_logger.go:
```diff
@@ -68,7 +68,7 @@ func (l *AuditLogger) logEvent(objMeta ObjectRef, gvk schema.GroupVersionKind, i
 			Name:        fmt.Sprintf("%v.%x", objMeta.Name, t.UnixNano()),
 			Labels:      eventLabels,
-			Annotations: logFields,
+			Annotations: eventAnnotations,
```
If we want to start adding information into logFields which won't make their way into the event, we need to split this parameter into two.

This is why the PR has exploded in size.

crenshaw-dev commented on server/application/application.go:
```diff
@@ -987,6 +993,23 @@ func (s *Server) updateApp(ctx context.Context, app *v1alpha1.Application, newAp
+	if err != nil {
+		return "", fmt.Errorf("failed to create json diff between applications: %w", err)
+	}
+	return string(patch), nil
```
Would it be possible to unmarshal the patch into a map[string]any so log tools aren't forced to parse a JSON string?

TomHellier committed:
Update the audit log to include metadata, and pass in map[string]any as logFields

Addressing code review comments as part of #23131, also added the error to the logFields for appcontroller to stop unparam lint. It makes sense to keep the logFields in that function call despite nothing using it at the moment, so may as well pass in the error string.

Converted the json string into a map[string]any as that is what sirupsen/logrus expects for the WithField api.

Also support the diff on updates to ApplicationSpecs.

Signed-off-by: Tom Hellier <me@tomhellier.com>

TomHellier commented on server/application/application.go:
```diff
@@ -956,6 +956,10 @@ func (s *Server) waitSync(app *v1alpha1.Application) {
 }
 
 func (s *Server) updateApp(ctx context.Context, app *v1alpha1.Application, newApp *v1alpha1.Application, merge bool) (*v1alpha1.Application, error) {
+	applicationPatch, err := diffBetweenApplicationSpecs(&app.Spec, &newApp.Spec)
```
I've added metadata, I didn't really care about it for my purposes but I think adding metadata is helpful as annotations might be set on the application which affects behaviour.

TomHellier commented:
Hey @crenshaw-dev, please could I get another review at your convenience, thanks.

TomHellier commented:
Hi @crenshaw-dev  - is there anything else you'd like for this?
